akka-stream/src/main/scala/akka/stream/impl/LazySourceAsync.scala

```diff
@@ -0,0 +1,107 @@
+/*
+ * Copyright (C) 2015-2019 Lightbend Inc. <https://www.lightbend.com>
+ */
+
+package akka.stream.impl
+
+import akka.annotation.InternalApi
+import akka.dispatch.ExecutionContexts
+import akka.stream._
+import akka.stream.impl.Stages.DefaultAttributes
+import akka.stream.scaladsl.{ Keep, Source }
+import akka.stream.stage._
+
+import scala.concurrent.{ Future, Promise }
+import scala.util.{ Failure, Success, Try }
+import scala.util.control.NonFatal
+
+/**
+ * INTERNAL API
+ */
+@InternalApi private[akka] final class LazySourceAsync[T, M](sourceFactory: () ⇒ Future[Source[T, M]]) extends GraphStageWithMaterializedValue[SourceShape[T], Future[Option[M]]] {
```

Comment:
I really think we should avoid another lazy stage here and try to achieve this through composition instead. Tricky with the materialized value though.

Reply:
An alternative is that this should be the _only_ lazy stage and we compose the other factories with it.

Reply:
```Source.lazily(() => Source.fromFutureSource())``` does the trick. We will get the matval semantics of `fromFutureSource` but I think that is better than the sink one as you can easily make your `Future[M]` into a `Future[Option[M]]` with `recover` and `map` but if you try to go the other way around there you will have to invent an exception.